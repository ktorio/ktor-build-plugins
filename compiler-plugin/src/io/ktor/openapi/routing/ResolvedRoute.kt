package io.ktor.openapi.routing

data class ResolvedRoute(
    val path: String,
    val method: String,
    val fields: RouteFieldList = emptyList(),
)