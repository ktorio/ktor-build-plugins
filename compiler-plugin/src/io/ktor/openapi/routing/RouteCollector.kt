package io.ktor.openapi.routing

import org.jetbrains.kotlin.fir.FirSession

object RouteCollector {
    fun collectRoutes(graph: RouteCallGraph): List<ResolvedRoute> {
        // Process each route node to generate ResolvedRoutes
        return graph.routes.flatMap { route ->
            if (route.method == null) return@flatMap emptyList()
            
            // Find all paths from roots to this route
            val allPaths = graph.findAllPathsToRoots(route)
            
            // Process each path and merge fields
            with(graph.session) {
                allPaths.mapNotNull { path ->
                    resolve(
                        parents = path,
                        children = graph.getChildren(route)
                    )
                }
            }
        }
    }

    context(session: FirSession)
    fun resolve(
        parents: Iterable<RouteNode> = emptyList(),
        children: Iterable<RouteNode> = emptyList(),
    ): ResolvedRoute? {
        var stack = RouteStack(session)
        var fields = emptyList<RouteField>()
        for (ancestor in parents) {
            fields = ancestor.fields(stack).merge(fields)
            stack += ancestor
        }

        for (child in children) {
            fields = fields.merge(child.fields(stack))
        }

        val path = fields.path ?: return null
        val method = fields.method ?: return null
        if (RouteField.Ignore in fields) return null

        return ResolvedRoute(
            path = path,
            method = method,
            fields = fields,
        )
    }
}