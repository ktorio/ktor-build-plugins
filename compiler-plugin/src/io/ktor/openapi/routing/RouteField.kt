package io.ktor.openapi.routing

import io.ktor.openapi.model.JsonSchema
import io.ktor.openapi.model.JsonType
import kotlinx.serialization.json.JsonElement
import org.jetbrains.kotlin.utils.addToStdlib.firstIsInstanceOrNull

/**
 * Sealed class representing different KDoc parameters for OpenAPI documentation.
 */
sealed interface RouteField {

    fun merge(other: RouteField): RouteField? =
        if (this::class == other::class) this else null

    sealed interface SchemaHolder : RouteField {
        val schema: SchemaReference?
        val attributes: Map<String, JsonElement>
    }

    sealed interface Content : SchemaHolder {
        val contentType: String?
        val description: String?
    }

    sealed interface Parameter : SchemaHolder {
        val name: String
        val description: String?
        val `in`: String
    }

    /**
     * Built into the OpenAPI JSON structure, so ignored when building the path item.
     */
    sealed interface Transient : RouteField

    data class Method(val method: String) : Transient {
        override fun merge(other: RouteField): RouteField? =
            if (other is Method && method == other.method) this else null
    }

    data class Path(val path: String) : Transient {
        override fun merge(other: RouteField): RouteField? =
            if (other is Path)
                Path("${other.path}/${path.removePrefix("/")}")
            else null
    }

    data class Schema(
        val name: String,
        val schema: JsonSchema
    ) : Transient {
        override fun merge(other: RouteField): RouteField? =
            if (other is Schema && other.name == name) this else null
    }

    /**
     * Associates the endpoint with a tag for grouping.
     *
     * Format: `@tag TagName`
     */
    data class Tag(val name: String) : RouteField {
        override fun merge(other: RouteField): RouteField? =
            if (other is Tag && name == other.name) this else null
    }

    /**
     * Describes a path parameter.
     *
     * Format: `@path name description`
     */
    data class PathParam(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "path"
        override fun merge(other: RouteField): RouteField? =
            if (other is PathParam && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Describes a query parameter.
     *
     * Format: `@query name description`
     */
    data class QueryParam(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "query"
        override fun merge(other: RouteField): RouteField? =
            if (other is QueryParam && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Describes a header parameter.
     *
     * Format: `@header name description`
     */
    data class Header(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "header"
        override fun merge(other: RouteField): RouteField? =
            if (other is Header && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
            ) else null
    }

    /**
     * Describes a cookie parameter.
     *
     * Format: `@cookie name description`
     */
    data class Cookie(
        override val name: String,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Parameter {
        override val `in`: String = "cookie"
        override fun merge(other: RouteField): RouteField? =
            if (other is Cookie && name == other.name) copy(
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Documents the request body type.
     *
     * Format: `@body [Type] description`
     */
    data class Body(
        override val contentType: String? = null,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Content {
        override fun merge(other: RouteField): RouteField? =
            if (other is Body) copy(
                contentType = contentType ?: other.contentType,
                schema = schema ?: other.schema,
                description = description ?: other.description,
                attributes = attributes + other.attributes,
            ) else null
    }

    /**
     * Documents a response code with optional type and description.
     *
     * Format: `@response code [Type] description`
     */
    data class Response(
        val code: String,
        override val contentType: String? = null,
        override val schema: SchemaReference? = null,
        override val description: String? = null,
        override val attributes: Map<String, JsonElement> = emptyMap(),
    ) : Content {
        override fun merge(other: RouteField): RouteField? =
            if (other is Response && code == other.code)
                copy(
                    contentType = contentType ?: other.contentType,
                    schema = schema ?: other.schema,
                    description = description ?: other.description,
                    attributes = attributes + other.attributes,
                )
            else null
    }

    /**
     * Marks an endpoint as deprecated.
     *
     * Format: `@deprecated reason`
     */
    data class Deprecated(val reason: String) : RouteField

    /**
     * Provides a detailed endpoint description.
     *
     * Format: `@description text`
     */
    data class Description(val text: String) : RouteField


    /**
     * Provides a summary of the endpoint.
     * Format: `@summary text`
     */
    data class Summary(val text: String) : RouteField

    /**
     * Documents security requirements.
     *
     * Format: `@security scheme`
     */
    data class Security(val scheme: String) : RouteField {
        override fun merge(other: RouteField): RouteField? =
            if (other is Security && scheme == other.scheme) this else null
    }
}

typealias RouteFieldList = List<RouteField>

val RouteFieldList.path: String?
    get() = firstIsInstanceOrNull<RouteField.Path>()?.path

val RouteFieldList.method: String?
    get() = firstIsInstanceOrNull<RouteField.Method>()?.method

/**
 * Merges two lists of route fields into a new list.
 */
fun RouteFieldList.merge(other: RouteFieldList) = buildList {
    val otherMutable = other.toMutableList()
    for (field in this@merge) {
        val match = otherMutable.indices.firstNotNullOfOrNull { i ->
            field.merge(otherMutable[i])?.also {
                otherMutable.removeAt(i)
                add(it)
            }
        }
        // if no compatible field was found, add this field to the result
        if (match == null) add(field)
    }
    // add all remaining unmerged fields to the result
    // TODO ignore all summary fields because we currently have some false positives
    addAll(otherMutable.filterNot { it is RouteField.Summary })
}

sealed interface SchemaReference {

    fun asSchema(): JsonSchema

    data class Resolved(
        val schema: JsonSchema
    ) : SchemaReference {
        override fun asSchema(): JsonSchema = schema
    }

    sealed interface Link: SchemaReference {
        val name: String

        data class Simple(override val name: String, val jsonType: JsonType) : Link {
            override fun asSchema(): JsonSchema =
                JsonSchema(jsonType)
        }
        data class Reference(override val name: String) : Link {
            override fun asSchema(): JsonSchema =
                JsonSchema(ref = "#/components/schemas/$name")
        }
        data class Array(val element: Link) : Link by element {
            override fun asSchema(): JsonSchema =
                JsonSchema(
                    type = JsonType.array,
                    items = element.asSchema()
                )
        }
        data class Optional(val delegate: Link) : Link by delegate {
            override fun asSchema(): JsonSchema =
                delegate.asSchema()
                    .copy(required = false)
        }
    }
}

fun SchemaReference.hasReference(): Boolean = getReference() != null

fun SchemaReference.getReference(): String? = when(this) {
    is SchemaReference.Resolved -> schema.ref?.substringAfterLast('/')
    is SchemaReference.Link.Array -> element.getReference()
    is SchemaReference.Link.Optional -> delegate.getReference()
    is SchemaReference.Link.Reference -> name
    is SchemaReference.Link.Simple -> null
}