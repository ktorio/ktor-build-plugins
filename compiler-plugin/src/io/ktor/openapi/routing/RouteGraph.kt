package io.ktor.openapi.routing

import org.jetbrains.kotlin.fir.FirSession

/**
 * Represents a graph of routes where a node can have multiple parents.
 */
class RouteGraph(val session: FirSession) {
    private val parentToChildren = mutableMapOf<RouteNode, MutableSet<RouteNode>>()
    private val childToParents = mutableMapOf<RouteNode, MutableSet<RouteNode>>()
    internal val routes = mutableListOf<RouteNode.Route>()
    internal val functions = mutableListOf<RouteNode.Function>()

    fun isEmpty() = routes.isEmpty()

    fun add(node: RouteNode) {
        for (i in routes.indices.reversed()) {
            val parent = routes[i]
            if (node in parent) {
                addEdge(parent, node)
                break
            } else if (parent.filePath != node.filePath || parent.isTopLevel()) {
                break
            }
        }
        when (node) {
            is RouteNode.Function -> functions.add(node)
            is RouteNode.Route -> routes.add(node)
            is RouteNode.CallFeature -> {}
        }
    }

    /**
     * Now that the function nodes are all accounted for, we can look for their children to
     * populate the edges.
     */
    fun build(): RouteGraph {
        for (function in functions) {
            for (other in functions) {
                if (other in function && other.isTopLevel())
                    addEdge(other, function)
            }
            for (route in routes) {
                if (route in function && route.isTopLevel())
                    addEdge(function, route)
            }
        }
        return this
    }

    fun RouteNode.isTopLevel(): Boolean =
        childToParents[this].orEmpty().none { it is RouteNode.Route }

    /**
     * Adds a parent-child relationship between nodes.
     */
    private fun addEdge(parent: RouteNode, child: RouteNode) {
        parentToChildren.getOrPut(parent) { mutableSetOf() }.add(child)
        childToParents.getOrPut(child) { mutableSetOf() }.add(parent)
    }

    /**
     * Checks if a node has any parents.
     */
    private fun hasParents(node: RouteNode): Boolean =
        childToParents.containsKey(node) && childToParents[node]?.isNotEmpty() == true

    /**
     * Gets all parents of a node.
     */
    private fun getParents(node: RouteNode): Set<RouteNode> =
        childToParents[node] ?: emptySet()

    /**
     * Gets all children of a node.
     */
    fun getChildren(node: RouteNode): Set<RouteNode> =
        parentToChildren[node] ?: emptySet()

    /**
     * Finds all paths from a node to root nodes (nodes with no parents).
     * Returns a list of paths, where each path is a list of nodes from the node to a root.
     */
    fun findAllPathsToRoots(node: RouteNode): List<List<RouteNode>> {
        if (!hasParents(node)) return listOf(listOf(node))

        val result = mutableListOf<List<RouteNode>>()
        val parents = getParents(node)

        for (parent in parents) {
            val parentPaths = findAllPathsToRoots(parent)
            for (path in parentPaths) {
                result.add(path + node)
            }
        }

        return result
    }
}