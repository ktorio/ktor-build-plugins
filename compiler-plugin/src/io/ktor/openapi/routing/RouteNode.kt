package io.ktor.openapi.routing

import io.ktor.compiler.utils.getArgument
import io.ktor.compiler.utils.getFunctionName
import io.ktor.compiler.utils.range
import org.jetbrains.kotlin.fir.expressions.FirFunctionCall
import org.jetbrains.kotlin.fir.expressions.FirLiteralExpression

/**
 * Code references to the Ktor routing API.  These recorded, then merged to form the model of the OpenAPI specification.
 */
sealed interface RouteNode: SourceCoordinates {
    val fir: FirFunctionCall
    val fields: RouteStack.() -> RouteFieldList

    val id: String get() = filePath.orEmpty() + startOffset.toString()
    val functionName: String get() = fir.getFunctionName()
    override val startOffset: Int get() = fir.source?.range?.first ?: -1
    override val endOffset: Int get() = fir.source?.range?.last ?: -1

    /**
     * Route call with method and path.
     *
     * These map to paths in the OpenAPI specification.
     */
    class Route(
        override val filePath: String?,
        override val fir: FirFunctionCall,
        override val fields: RouteStack.() -> RouteFieldList,
    ): RouteNode {
        val method: String? get() =
            functionName.takeIf { it in RoutingFunctionConstants.HTTP_METHODS }

        override fun toString(): String = buildString {
            append(fir.getFunctionName())
            (fir.getArgument("path") as? FirLiteralExpression)?.let {
                append(" ")
                append(it.value.toString())
            }
        }

        override fun equals(other: Any?): Boolean =
            other is Route && id == other.id

        override fun hashCode(): Int = id.hashCode()
    }

    /**
     * A general call to some Ktor feature that generates automatic OpenAPI properties.
     */
    class CallFeature(
        override val filePath: String?,
        override val fir: FirFunctionCall,
        override val fields: RouteStack.() -> RouteFieldList,
    ) : RouteNode {
        override fun equals(other: Any?): Boolean =
            other is CallFeature && id == other.id

        override fun hashCode(): Int = id.hashCode()
    }

    /**
     * Custom extension function of `Route`.
     *
     * Since it's a reference to a function, it may have non-local children.
     */
    class Function(
        override val filePath: String?,
        override val fir: FirFunctionCall,
        val declaration: SourceCoordinates,
        override val fields: RouteStack.() -> RouteFieldList,
    ): RouteNode {
        override fun contains(other: SourceCoordinates): Boolean =
            other in declaration && other != this

        override fun toString(): String =
            fir.getFunctionName() + '@' + fir.source?.range

        override fun equals(other: Any?): Boolean =
            other is Function && id == other.id

        override fun hashCode(): Int = id.hashCode()
    }
}