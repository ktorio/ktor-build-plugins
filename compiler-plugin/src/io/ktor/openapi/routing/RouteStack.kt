package io.ktor.openapi.routing

import io.ktor.compiler.utils.*
import org.jetbrains.kotlin.fir.FirSession
import org.jetbrains.kotlin.fir.expressions.FirExpression

data class RouteStack(
    val session: FirSession,
    val stack: List<RouteNode> = emptyList(),
) {
    operator fun plus(node: RouteNode): RouteStack =
        RouteStack(session, stack + node)

    fun FirExpression.evaluate() =
        with(session) {
            createCallStackEvaluator(stack.map { it.fir })
        }.evaluate(this)
}