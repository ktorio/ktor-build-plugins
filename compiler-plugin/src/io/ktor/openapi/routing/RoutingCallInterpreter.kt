package io.ktor.openapi.routing

import io.ktor.openapi.model.JsonSchema
import kotlinx.serialization.Serializable
import org.jetbrains.kotlin.diagnostics.DiagnosticReporter
import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
import org.jetbrains.kotlin.fir.expressions.FirFunctionCall

/**
 * Infers aspects for OpenAPI generation from a general function call in the source code.
 */
fun interface RoutingCallInterpreter {
    context(context: CheckerContext, reporter: DiagnosticReporter)
    fun check(expression: FirFunctionCall): RoutingReferenceResult
}

sealed interface RoutingReferenceResult {
    /**
     * Nothing found for this function call.
     */
    object None : RoutingReferenceResult

    /**
     * Route node found.
     */
    data class Match(val call: RouteNode) : RoutingReferenceResult

    /**
     * `install(ContentNegotiation)`
     */
    data class ContentType(val contentType: io.ktor.openapi.routing.ContentType): RoutingReferenceResult

    /**
     * `authentication {}`
     */
    data class SecurityScheme(
        val name: String,
        val type: String,
        val scheme: String? = null,
        val bearerFormat: String? = null,
        val openIdConnectUrl: String? = null,
        val flows: Map<String, OauthFlow>? = null,
    ) : RoutingReferenceResult
}

@Serializable
data class OauthFlow(
    val authorizationUrl: String,
    val tokenUrl: String,
    val refreshUrl: String? = null,
    val scopes: Map<String, String>? = null,
)