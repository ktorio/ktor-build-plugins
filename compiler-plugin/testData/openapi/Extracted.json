{
    "openapi": "3.1.1",
    "info": {
        "title": "Open API Document",
        "version": "1.0.0"
    },
    "paths": {
        "/api/users": {
            "post": {
                "summary": "Save a new user.\n",
                "requestBody": {
                    "description": "the user to save.",
                    "content": {
                        "application/json": {
                            "schema": {
                                "type": "[User1]"
                            }
                        }
                    }
                }
            },
            "get": {
                "summary": "Get a list of users.\n",
                "responses": {
                    "200": {
                        "description": "A list of users.",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "[List<User>]"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/api/users/{id}": {
            "delete": {
                "summary": "Delete a user.\n",
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "The ID of the user",
                        "required": true
                    }
                ]
            },
            "get": {
                "summary": "Get a single user\n",
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "The ID of the user",
                        "required": true
                    }
                ],
                "responses": {
                    "404": {
                        "description": "The user was not found"
                    }
                }
            }
        }
    }
}