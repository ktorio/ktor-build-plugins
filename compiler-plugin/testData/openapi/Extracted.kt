// RUN_PIPELINE_TILL: BACKEND

package openapi

import io.ktor.http.HttpStatusCode
import io.ktor.serialization.kotlinx.json.json
import io.ktor.server.application.*
import io.ktor.server.plugins.contentnegotiation.ContentNegotiation
import io.ktor.server.request.receive
import io.ktor.server.routing.*
import io.ktor.server.response.*
import io.ktor.util.toMap

fun Application.extracted(repository: Repository1<User1>) {
    install(ContentNegotiation) {
        json()
    }

    routing {
        route("/api") {
            userEndpoints(repository)
        }
    }
}

private fun Route.userEndpoints(repository: Repository1<User1>) {
    route("/users") {
        readEndpoints(repository)
        modificationEndpoints(repository)
    }
}

private fun Route.modificationEndpoints(repository: Repository1<User1>) {
    /**
     * Save a new user.
     *
     * @body [User1] The user to save.
     */
    post {
        repository.save(call.receive())
        call.respond(HttpStatusCode.Created)
    }

    /**
     * Delete a user.
     * @path id The ID of the user
     */
    delete("{id}") {
        repository.delete(call.parameters["id"]!!)
        call.respond(HttpStatusCode.NoContent)
    }
}

private fun Route.readEndpoints(repository: Repository1<User1>) {
    /**
     * Get a list of users.
     *
     * @response 200 [User1]+ A list of users.
     */
    get {
        val query = call.request.queryParameters.toMap()
        val list = repository.list(query)
        call.respond(list)
    }

    /**
     * Get a single user
     *
     * @path id The ID of the user
     * @response 404 The user was not found
     */
    get("{id}") {
        val user = repository.get(call.parameters["id"]!!)
            ?: return@get call.respond(HttpStatusCode.NotFound)
        call.respond(user)
    }
}

interface Repository1<E> {
    fun get(id: String): E?
    fun save(entity: E)
    fun delete(id: String)
    fun list(query: Map<String, List<String>>): List<E>
}

data class User1(val id: String, val name: String)