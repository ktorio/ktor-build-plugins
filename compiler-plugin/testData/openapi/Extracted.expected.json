{
    "openapi": "3.1.1",
    "info": {
        "title": "Open API Document",
        "version": "1.0.0"
    },
    "paths": {
        "/api/users": {
            "post": {
                "summary": "Save a new user.",
                "requestBody": {
                    "description": "The user to save.",
                    "content": {
                        "application/json": {
                            "schema": {
                                "$ref": "#/components/schemas/User1"
                            }
                        }
                    }
                }
            },
            "get": {
                "summary": "Get a list of users.",
                "responses": {
                    "200": {
                        "description": "A list of users.",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "array",
                                    "items": {
                                        "$ref": "#/components/schemas/User1"
                                    }
                                }
                            }
                        }
                    }
                },
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "The ID of the user",
                        "required": true,
                        "schema": {
                            "type": "string"
                        }
                    }
                ]
            }
        },
        "/api/users/{id}": {
            "delete": {
                "summary": "Delete a user.",
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "The ID of the user",
                        "required": true,
                        "schema": {
                            "type": "string"
                        }
                    }
                ]
            },
            "get": {
                "summary": "Get a single user",
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "The ID of the user",
                        "required": true,
                        "schema": {
                            "type": "string"
                        }
                    }
                ],
                "responses": {
                    "404": {
                        "description": "The user was not found"
                    },
                    "200": {
                        "description": "",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/User1"
                                }
                            }
                        }
                    }
                }
            }
        },
        "/api/messages": {
            "post": {
                "summary": "Save a new message.",
                "requestBody": {
                    "description": "The message to save.",
                    "content": {
                        "application/json": {
                            "schema": {
                                "$ref": "#/components/schemas/Message1"
                            }
                        }
                    }
                },
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "The ID of the user",
                        "required": true,
                        "schema": {
                            "type": "string"
                        }
                    }
                ],
                "responses": {
                    "404": {
                        "description": "The user was not found"
                    }
                }
            },
            "get": {
                "summary": "Get a list of messages.",
                "responses": {
                    "200": {
                        "description": "A list of messages.",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "array",
                                    "items": {
                                        "$ref": "#/components/schemas/Message1"
                                    }
                                }
                            }
                        }
                    }
                },
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "The ID of the message",
                        "required": true,
                        "schema": {
                            "type": "string"
                        }
                    }
                ]
            }
        },
        "/api/messages/{id}": {
            "delete": {
                "summary": "Delete a message.",
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "The ID of the message",
                        "required": true,
                        "schema": {
                            "type": "string"
                        }
                    }
                ],
                "responses": {
                    "404": {
                        "description": "The user was not found"
                    }
                }
            },
            "get": {
                "summary": "Get a single message",
                "parameters": [
                    {
                        "name": "id",
                        "in": "path",
                        "description": "The ID of the message",
                        "required": true,
                        "schema": {
                            "type": "string"
                        }
                    }
                ],
                "responses": {
                    "404": {
                        "description": "The message was not found"
                    },
                    "200": {
                        "description": "",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/Message1"
                                }
                            }
                        }
                    }
                }
            }
        }
    },
    "components": {
        "schemas": {
            "User1": {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "string"
                    },
                    "name": {
                        "type": "string"
                    }
                }
            },
            "Message1": {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "string"
                    },
                    "text": {
                        "type": "string"
                    }
                }
            }
        }
    }
}